src/effects/backends/effectprocessor.h

```diff
@@ -148,6 +155,20 @@ class EffectProcessorImpl : public EffectProcessor {
         return 0;
     }
 
+    std::unique_ptr<QDialog> createUI() override {
+        QLabel* label = new QLabel();
+        label->setText("This effect has no UI");
+
+        QVBoxLayout* layout = new QVBoxLayout();
+        layout->addWidget(label);
+
+        std::unique_ptr<QDialog> dialog = std::make_unique<QDialog>();
+        dialog->setWindowTitle("No UI available");
+        dialog->setLayout(layout);
+
+        return dialog;
+    }
```

Comment:
is this really needed? This should only be shown when a plugin has `hasUI` is true even though it shouldn't be right?

Reply:
Exactly, I haven't gotten around to do that check yet and decided that this was nicer than an empty dialog 😄 

I've also considered dropping the `hasUI` and just returning an optional dialog here, but that wouldn't give the UI a chance to hide/style the UI button beforehand.

Reply:
yeah... I think hiding the button if not available would be definitely good to have. I'd prefer to keep that if we can.

Reply:
One consequence is that we keep a dialog open when the user switches to an effect that doesn't provide a UI. On the one hand I think this could be nice (e.g. the user might arrange their effect UIs on the screen in a particular way and it might be nice to keep them open until explicitly closed). On the other hand, the blanket implementation doesn't really do anything useful, so we might as well also just close the UI, in which case the next effect with a UI will automatically open the dialog since the control object enabling the UI is still toggled on.

Don't really have a preference what the best/most intuitive behavior is here. Any thoughts?

Reply:
I think we should follow the approach DAWs usually take. You can open as many effect UIs as there are effects and if you replace one effect with one that doesn't have a window or you remove the effect, the window just closes. Now you usually don't replace effects in DAWs directly (you insert a new one and remove the old one) but in mixxx we can replace an effect directly since we have slots. In that case we can consider replacing the old UI with the new UI directly (if the old effect had its UI open). Wdyt?

Reply:
> Now you usually don't replace effects in DAWs directly

Depends on the DAW, at least in Logic the workflow is similar to Mixxx, where effects are essentially also slots with a dropdown.

> in mixxx we can replace an effect directly since we have slots. In that case we can consider replacing the old UI with the new UI directly (if the old effect had its UI open). Wdyt?

Sounds reasonable. The way the API is designed currently, we'd still instantiate a new dialog (the new effect UI might want to set its own title/resizability/etc.), but I was thinking of maybe just carrying over the position manually.

Reply:
> Depends on the DAW, at least in Logic the workflow is similar to Mixxx, where effects are essentially also slots with a dropdown.

ah yeah. My primary experience lies with bitwig and ableton... 

> but I was thinking of maybe just carrying over the position manually.

That may not be feasible on all platforms (not particularly relevant for MacOS I guess, but I don't think we could implement the same on wayland for example).